test/fixtures/es-module-loaders/hooks-input.mjs

```diff
@@ -0,0 +1,92 @@
+// This is expected to be used by test-esm-loader-hooks.mjs via:
+// node --loader ./test/fixtures/es-module-loaders/hooks-input.mjs ./test/fixtures/es-modules/json-modules.mjs
+
+import assert from 'assert';
+import { write } from 'fs';
+import { readFile } from 'fs/promises';
+import { fileURLToPath } from 'url';
+
+
+let resolveCalls = 0;
+let loadCalls = 0;
+
+export async function resolve(specifier, context, next) {
+  resolveCalls++;
+  let url;
+
+  if (resolveCalls === 1) {
+    url = new URL(specifier).href;
+    assert.match(specifier, /json-modules\.mjs$/);
+    assert.deepStrictEqual(context.parentURL, undefined);
+    assert.deepStrictEqual(context.importAssertions, {
+      __proto__: null,
+    });
+  } else if (resolveCalls === 2) {
+    url = new URL(specifier, context.parentURL).href;
+    assert.match(specifier, /experimental\.json$/);
+    assert.match(context.parentURL, /json-modules\.mjs$/);
+    assert.deepStrictEqual(context.importAssertions, {
+      __proto__: null,
+      type: 'json',
+    });
+  }
+
+  // Ensure `context` has all and only the properties it's supposed to
+  assert.deepStrictEqual(Object.keys(context), [
+    'conditions',
+    'importAssertions',
+    'parentURL',
+  ]);
+  assert.ok(Array.isArray(context.conditions));
+  assert.deepStrictEqual(typeof next, 'function');
+
+  const returnValue = {
+    url,
+    format: 'test',
+    shortCircuit: true,
+  }
+
+  await new Promise(resolve => write(1, `${JSON.stringify(returnValue)}\n`, resolve)); // For the test to read
```

Comment:
I don't understand the comment, I think we need to explain in more details what this `new Promise(() => write(1))` thing is, and we should probably add a TODO comment to use something else, something that users are actually likely to write.

Reply:
I guess for this version it could be simple `console.log`, but once this is inside the worker thread `console.log` doesn’t print to stdout anymore and you need constructs like this.

Reply:
I meant a comment in the code, I am aware why you wrote it like this, future readers might not be.

Reply:
I think I’ll update from `console.log` to `fs.writeSync` or similar for all of the fixture loaders together as part of the off-thread PR, I think that makes more sense. How about this?

```suggestion
  console.log(JSON.stringify(returnValue)); // For the test to validate when it parses stdout
```

Reply:
Yeah that makes sense, and in the off-thread PR you could use `globalPreload` to "polyfill" `console.log` to the hack you found.

Reply:
The hack is needed for loaders themselves, as they’re the ones running inside the thread. I don’t have to use the promise version, I could just do ``writeSync(1, `${message}\n`); // Log to stdout`` which is probably simple enough. Though it feels like `fs.write` should have a counterpart in `fs/promises`.

Reply:
I meant doing something like:

```js
const console = {};
export function globalPreload({ port }) {
  console.log = (...args) => port.postMessage(args);
  return 'port.onmessage = evt => console.log(...evt.data)';
}
```

Reply:
As in, pass the message to the main thread via application code rather than to the main thread via Node internals? That’s a creative solution, but I would think that it only works if the application code runs without errors. If a loader is trying to print an error like “source failed transpilation” or somesuch, we can’t wait until the program entry point has successfully parsed and started executing. Even if somehow the sloppy script still executes if the user code doesn’t, it still feels really complicated where there are probably edge cases where it fails and `writeSync` wouldn’t.

Reply:
I just tried it and you're incorrect, the code runs no matter if the entry point is successfully parsed or not:

```js
/// loader.mjs
const console = {};
export function globalPreload({ port }) {
  console.log = function log() { port.postMessage([...arguments]); };
  return 'port.onmessage = evt => console.log(...evt.data)';
}

export async function resolve(url, context, next) {
  console.log('logging from loader thread', url);
  return next(url, context);
}
```

```console
$ node --loader ./loader.mjs test/fixtures/syntax/bad_syntax.mjs 
(node:17727) ExperimentalWarning: Custom ESM Loaders is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
logging from loader thread file://…/test/fixtures/syntax/bad_syntax.mjs
file://…/test/fixtures/syntax/bad_syntax.mjs:1
var foo bar;
        ^^^

SyntaxError: Unexpected identifier 'bar'
    at ESMLoader.moduleStrategy (node:internal/modules/esm/translators:117:18)
    at ESMLoader.moduleProvider (node:internal/modules/esm/loader:459:14)
    at async link (node:internal/modules/esm/module_job:68:21)

Node.js v19.3.0
```

FWIW I don't think `writeSync(1)` passes anything to the main thread AFAIK, it passes to the FS which outputs it on the stdout (unless libuv treats fd of `1` differently maybe?)

Reply:
> FWIW I don’t think `writeSync(1)` passes anything to the main thread AFAIK, it passes to the FS which outputs it on the stdout (unless libuv treats fd of `1` differently maybe?)

Good point. I guess the question is, if we want to polyfill `console.log` within a worker thread, why rely on `globalPreload` when we could do something like this:

```js
console.log = (...args) => require('fs').writeSync(1, args.map(arg => `${require('util').inspect(arg)}\n`).join(' '));
```

Reply:
IMO user expectations of folks using `console.log` is to use `console.log`, not do a synchronous write operation on a file descriptor.

Reply:
Well yes; and they also expect `console.log` to print to stdout. But inside worker threads, it doesn't. I feel like we should fix that.